### ISSUE TITLE
How to call SATSPLN in refpropm.m flie

### ISSUE THREAD

**Issue by RayleighSun** [usnistgov/REFPROP-wrappers#436]
Hi, Dr. Bell and Dr. Lemmon,
I am the one who asked issue #434, and **_I've solved my problem both in Excel and MATLAB via Python interface_**, as I replied in  issue #434.
However, the problem still exsits in MATLAB via [`refpropm.m`,](https://github.com/usnistgov/REFPROP-wrappers/blob/master/wrappers/MATLAB/legacy/refpropm.m) as shown below:
```
>> refpropm('C','T',338.15,'P',8*1000,'CO2','propane',[0.5 1-0.5])/1000
Error using refpropm (line 519)
[TPFL2 error 226] 2-phase iteration did not converge. 
```
![image](https://user-images.githubusercontent.com/94279628/148671047-aaa8d6ba-77eb-486f-8d9b-dc6800bc32e3.png)

This is how I solve it in MATLAB via Python:
**_Befrore:_**
```
hFld = 'CO2;Propane';
RP = py.ctREFPROP.ctREFPROP.REFPROPFunctionLibrary('C:\Program Files (x86)\REFPROP');
iUnit = RP.MASS_BASE_SI;
iMass = int8(0); % 0: molar fractions; 1: mass fractions
iFlag = int8(0); % 0: don't call SATSPLN; 1: call SATSPLN
RP.SETFLUIDSdll(hFld);
z = {0.5,0.5};
r1 = RP.REFPROPdll(' ','TP','CP',iUnit,iMass,iFlag,338.15,8*1e6,z);
o1 = double(r1.Output);
cp = o1(1)/1e3

cp =

  -1.0000e+04
```
**_After:_**
```
hFld = 'CO2;Propane';
RP = py.ctREFPROP.ctREFPROP.REFPROPFunctionLibrary('C:\Program Files (x86)\REFPROP');
iUnit = RP.MASS_BASE_SI;
iMass = int8(0); % 0: molar fractions; 1: mass fractions
iFlag = int8(1); % 0: don't call SATSPLN; 1: call SATSPLN
RP.SETFLUIDSdll(hFld);
z = {0.5,0.5};
r1 = RP.REFPROPdll(' ','TP','CP',iUnit,iMass,iFlag,338.15,8*1e6,z);
o1 = double(r1.Output);
cp = o1(1)/1e3

cp =

    4.8351
```
According to the experience of solving the problem in Excal and Python that same state points do not converge by calling `SATSPLN`, I tried to uncomment the lines of 311 and 312 in `refpropm.m`
![image](https://user-images.githubusercontent.com/94279628/148671003-fd548d01-3488-4573-b494-581fe025c7ad.png)
but the problem seems not be solved. 
I am not sure if I need to move the lines of 311 and 312 to another place?

I'm aware that `refpropm.m` is deprecated, however as a `refpropm.m` long-term user, the truth is this m-flie is very fast in calculation and very easy to use, therefore I still hope to get help from you and the team.

**Comment by EricLemmon** [comment]
Glad to hear that at least things are partially working!

Does the value of ierr come back as zero at this point, and is herr a blank string?  Is the call to TPFL2 made right after the call to SATSPLN?  Also, is the call to SATSPLN located after the call to SETUP?

**Comment by ianhbell** [comment]
As you note, refpropm is fully deprecated now.  Is there a computational speed difference between refpropm and the Python interface?

**Comment by RayleighSun** [comment]
Thansk, Dr. Bell, I tried to compare computational speed difference between refpropm and the Python interface in MATLAB use "Run and Time". For single point or few point calculation, the computational speed difference is not obvious. 
But when I try to calculate the heat capacity, density and enthalpy as a function of temperature at three different pressures, the calculation speed is quite different, when I calculate CO2 + propane in CO2 mass fraction of 0.75 (_note that I used the mass fraction, temperature and pressure that `refpropm.m` can converge_). Here is the cost time results 
```
Elapsed time is 2.606023 seconds.
Elapsed time is 0.243600 seconds.
```
High-Level API cost 2.6 seconds and refpropm cost 0.25 seconds.
> 

![image](https://user-images.githubusercontent.com/94279628/148709230-ff8e3493-ca83-47ef-8639-79d19da76876.png)

Obviously, the most time-consuming part of the High-Level API program is converting the data type to double (line 23), followed by calling REFPROPdll (line 22). 

This computational speed difference becomes more significant when I try to use optimization algorithm and SIMULINK, and is the very reason I still struggle on trying the `refpropm.m`

Here is my demo code:
```
clear;clc
%% High-Level API
tic
w = 0.75; % CO2 mass fraction now
temperature = 320:1:420;
pressure = [10 15 20];
hFld = 'CO2;Propane';
RP = py.ctREFPROP.ctREFPROP.REFPROPFunctionLibrary('C:\Program Files (x86)\REFPROP');
iUnit = RP.MASS_BASE_SI;
iMass = int8(1); % 0: molar fractions; 1: mass fractions
iFlag = int8(1); % 0: don't call SATSPLN; 1: call SATSPLN
RP.SETFLUIDSdll(hFld); % must use, or bug
z = {w, 1-w};
cp = zeros(length(temperature),length(pressure));
D = zeros(length(temperature),length(pressure));
h = zeros(length(temperature),length(pressure));
for kk=1:length(pressure)
    for jj = 1:length(temperature)
        
        t = temperature(jj);
        p = pressure(kk);
        r1 = RP.REFPROPdll(' ','TP','CP;D;H',iUnit,iMass,iFlag,t,p*1e6,z);
        o1 = double(r1.Output);
        cp(jj, kk) = o1(1)/1e3;
        D(jj, kk) = o1(2);
        h(jj, kk) = o1(3)/1e3;
        
    end
end
toc

%% refpropm
tic
cp2 = zeros(length(temperature),length(pressure));
D2 = zeros(length(temperature),length(pressure));
h2 = zeros(length(temperature),length(pressure));
for kk = 1:length(pressure)
    for jj = 1:length(temperature)
        
        t = temperature(jj);
        p = pressure(kk);
        cp2(jj,kk) = refpropm('C','T',t,'P',p*1e3,'CO2','propane',[w, 1-w])/1e3;
        D2(jj,kk) = refpropm('D','T',t,'P',p*1e3,'CO2','propane',[w, 1-w]);
        h2(jj,kk) = refpropm('H','T',t,'P',p*1e3,'CO2','propane',[w, 1-w])/1e3;
        
    end
end

toc
```

**Comment by RayleighSun** [comment]
> Glad to hear that at least things are partially working!
> 
> Does the value of ierr come back as zero at this point, and is herr a blank string? Is the call to TPFL2 made right after the call to SATSPLN? Also, is the call to SATSPLN located after the call to SETUP?

Thanks, Dr. Lemmon, I'm trying according to your suggestion, I'll get back to you when I have the result.

**Comment by RayleighSun** [comment]
> Glad to hear that at least things are partially working!
> 
> Does the value of ierr come back as zero at this point, and is herr a blank string? Is the call to TPFL2 made right after the call to SATSPLN? Also, is the call to SATSPLN located after the call to SETUP?

Thanks, Dr. Lemmon, I think I solved it in `refpropm.m` now. I removed subroutine SATSPLN after I get the results of mole farction by XMOLEdll, as shown below:

![image](https://user-images.githubusercontent.com/94279628/149246290-a262a718-442b-48ad-9f34-5b6e9aa4a703.png)

Now the unconverged state points I have encountered can be calculated normally by `refpropm.m`, and it works well.

More importantly, I began to understand the calculation logic of the REFPROP eventually in different interface, thank you @EricLemmon  and Dr. Bell @ianhbell for your efforts.
